editoast/src/core/mq_client.rs

```diff
@@ -67,28 +192,102 @@ pub struct MQResponse {
 const SINGLE_WORKER_KEY: &str = "all";
 
 impl RabbitMQClient {
-    pub async fn new(options: Options) -> Result<Self, Error> {
+    pub async fn new(options: Options) -> Result<Self, MqClientError> {
         let hostname = hostname::get()
             .map(|name| name.to_string_lossy().into_owned())
             .unwrap_or_else(|_| "unknown".to_string());
 
-        let cfg = Config {
-            url: Some(options.uri),
-            ..Default::default()
-        };
-        let pool = cfg
-            .create_pool(Some(Runtime::Tokio1))
-            .map_err(Error::CreatePoolLapin)?;
+        let conn = Arc::new(RwLock::new(None));
+
+        tokio::spawn(Self::connection_loop(
+            options.uri,
+            hostname.clone(),
+            conn.clone(),
+        ));
+
+        // We should ensure that the connection is established at least once before creating the pool
+        // since deadpool will try to create resources upfront
+        const MAX_RETRIES: usize = 10;
+        const RETRY_DELAY: Duration = Duration::from_secs(1);
+
+        let mut retries = 0;
+        while retries < MAX_RETRIES {
+            if Self::connection_ok(&conn).await {
+                break;
+            }
+            tokio::time::sleep(RETRY_DELAY).await;
+            retries += 1;
+        }
+
+        if retries == MAX_RETRIES {
+            return Err(MqClientError::ConnectionDoesNotExist);
+        }
+
+        // Create the pool
+        let pool = Pool::builder(ChannelManager::new(conn, hostname))
+            .max_size(options.num_channels)
+            .build()
+            .map_err(|_| MqClientError::ConnectionDoesNotExist)?;
 
         Ok(RabbitMQClient {
             pool,
             exchange: format!("{}-req-xchg", options.worker_pool_identifier),
             timeout: options.timeout,
-            hostname,
             single_worker: options.single_worker,
         })
     }
 
+    async fn connection_ok(connection: &Arc<RwLock<Option<Connection>>>) -> bool {
+        let guard = connection.as_ref().read().await;
+        let conn = guard.as_ref();
+        let status = match conn {
+            None => return false,
+            Some(conn) => conn.status().state(),
+        };
+        match status {
+            lapin::ConnectionState::Initial => true,
+            lapin::ConnectionState::Connecting => true,
+            lapin::ConnectionState::Connected => true,
+            lapin::ConnectionState::Closing => true,
```

Comment:
Looks weird to consider both `Initial`,  `Connecting` and `Closing` as `true`. It feels that either we consider, it will eventually be ok, and in this case, `Closing` should be `false`, or we want to consider the connection can now take some traffic, and in this case, `Initial` and `Connecting` should be false. What do you think?

Reply:
Agree on this, we should only consider the connected state.